File: bin/xkcd_password.py
#!/usr/bin/env python

# xkcd_password.py
# Simple python script to generate an xkcd password
# Recommended in http://xkcd.com/936/

import random
import re
import sys
import urllib2

def numberizeVowels(word):
    changed = ''
    for letter in word:
        if letter == 'a':
            changed += '4'
        elif letter == 'e':
            changed += '3'
        elif letter == 'i':
            changed += '1'
        elif letter == 'o':
            changed += '0'
        elif letter == 'u':
            changed += '2'
        else:
            changed += letter
    return changed

def generate(word_count = 4, joiner = '-', password_length = None):
    dictionaryURL = 'http://scrapmaker.com/data/wordlists/twelve-dicts/2of12.txt';
    randwords = []
    data = urllib2.urlopen(dictionaryURL)
    lines = data.read().split()
    generated = ''
    first = True
    for i in xrange(word_count):
        word = lines[random.randint(0, len(lines) - 1)].strip()

        # If the word is not all lowercase letters, get a new one
        while(re.match('^[a-z]+$', word) == None):
            word = lines[random.randint(0, len(lines) - 1)].strip()

        # Numberize first word
        if first:
            word = numberizeVowels(word)
            first = False

        # append to password
        randwords.append(word)
        generated += word.capitalize()

    if password_length != None and len(generated) > password_length:
        generated = generated[:password_length]

    return generated

if __name__ == '__main__':
    if len(sys.argv) == 2:
        print generate(password_length=int(sys.argv[1]))
    else:
        print generate()

File: bin/xkcd_password.js
#!/usr/bin/env node

// Simple Node.js script to generate an xkcd password
// Based on recommendation in http://xkcd.com/936/

var crypto = require('crypto');
var fs = require('fs');
var path = require('path');

// Configuration
var dir = path.dirname(__filename);
var dictionaryFile = path.join(dir, '../conf/2of12.txt');
var phraseCount = 4;

function numberizeVowels(word) {
  var changed = '';
  var map = {
    a: 4,
    e: 3,
    i: 1,
    o: 0,
    u: 2
  }
  for (var i = 0; i < word.length; i++) {
    var letter = word[i];
    if (map.hasOwnProperty(word[i])) {
      changed += map[word[i]];
    } else {
      changed += letter;
    }
  }
  return changed;
}

// Returns a secure random int in the range [min, max)
function randInt(min, max) {
  var buf = crypto.randomBytes(32);
  var rand = Math.abs(buf.readInt32BE(0));
  var spread = max - min;
  return rand % spread + min;
}

function generate(length) {
  var data = fs.readFileSync(dictionaryFile).toString();
  var lines = data.split('\n');
  generated = '';
  first = true;

  for (var i = 0; i < phraseCount; i++) {
    var word = lines[randInt(0, lines.length)].trim();

    // If the word is not all lowercase letters, get a new one
    while (word.match(/^[a-z]+$/) === null) {
      word = lines[randInt(0, lines.length)].trim();
    }

    // Capitalize
    word = word.charAt(0).toUpperCase() + word.slice(1);

    // Numberize first word
    if (first) {
      word = numberizeVowels(word);
      first = false;
    }

    // Append to password
    generated += word;
  }

  if (typeof length != 'undefined' && generated.length > length) {
    generated = generated.substring(0, length);
  }

  return generated;
}

if (require.main === module) {
  if (process.argv.length > 2) {
    console.log(generate(parseInt(process.argv[2])));
  } else {
    console.log(generate());
  }
}
